File: nanopay/src/net/nanopay/tx/BlockDisabledUserTransactionDAO.java
package net.nanopay.tx;

import foam.core.FObject;
import foam.core.X;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.nanos.auth.User;
import net.nanopay.tx.model.Transaction;

public class BlockDisabledUserTransactionDAO extends ProxyDAO {
  private static final String BLOCK_TRANSACTION = "Transaction: %d is blocked because the %s is disabled.";

  public BlockDisabledUserTransactionDAO(X x, DAO delegate) {
    super(x, delegate);
  }

  @Override
  public FObject put_(X x, FObject obj) {
    Transaction txn = (Transaction) obj;
    User payer = txn.findSourceAccount(x).findOwner(x);
    User payee = txn.findDestinationAccount(x).findOwner(x);

    if (!payer.getEnabled()) {
      blockTransaction(txn, "payer");
    }

    if (!payee.getEnabled()) {
      blockTransaction(txn, "payee");
    }

    return super.put_(x, obj);
  }

  private void blockTransaction(Transaction transaction, String user) {
    throw new RuntimeException(
      String.format(BLOCK_TRANSACTION, transaction.getId(), user));
  }
}

File: nanopay/src/net/nanopay/meter/BlockDisabledUserTransactionDAO.js
foam.CLASS({
  package: 'net.nanopay.meter',
  name: 'BlockDisabledUserTransactionDAO',
  extends: 'foam.dao.ProxyDAO',

  documentation: `Decorating DAO for preventing creation of transaction
      involving a disabled user as a payer or payee.`,

  javaImports: [
    'foam.dao.DAO',
    'foam.nanos.auth.User',
    'net.nanopay.account.Account',
    'net.nanopay.tx.model.Transaction'
  ],

  constants: [
    {
      name: 'BLOCK_TRANSACTION',
      type: 'String',
      value: 'Transaction: %s is blocked because the %s is disabled.'
    }
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        Transaction txn = (Transaction) obj;

        if (!checkAccountOwner(x, txn.findSourceAccount(x))) {
          blockTransaction(txn, "payer");
        }

        if (!checkAccountOwner(x, txn.findDestinationAccount(x))) {
          blockTransaction(txn, "payer");
        }

        return super.put_(x, obj);
      `
    },
    {
      name: 'checkAccountOwner',
      javaReturns: 'boolean',
      args: [
        { of: 'foam.core.X', name: 'x' },
        { of: 'Account', name:'account' }
      ],
      javaCode: `
        if (account != null) {
          User user = account.findOwner(x);

          if (user != null) {
            return user.getEnabled();
          }
        }
        return true;
      `
    },
    {
      name: 'blockTransaction',
      javaReturns: 'void',
      args: [
        { of: 'Transaction', name: 'transaction' },
        { of: 'String', name: 'user' }
      ],
      javaCode: `
        throw new RuntimeException(
          String.format(BLOCK_TRANSACTION, transaction.getId(), user));
      `
    }
  ]
});
